### ISSUE TITLE
Migrate the Effect-TS codebase to mirror ZIO 2.x

### ISSUE THREAD

**Issue by IMax153** [Effect-TS/core#1035]
The [`feat/system-next`](https://github.com/Effect-TS/core/tree/feat/system-next) branch is a work-in-progress port of the changes introduced in ZIO `v2.x` to Effect-TS. This is very early work and is by no means feature-complete - many features are missing, do not work properly, or are implemented inefficiently. However, the basic ability to run effects has been built.

I would like to ask the rest of the Effect-TS community to help continue the work started on this branch. The plan will be that individuals who would like to contribute to this effort should select something either from the list below (or an unlisted feature that they would like to port from ZIO 2.x) and state that they will be working on that feature in the comments. This way we don't have efforts being duplicated. Pull requests should be opened against the `feat/system-next` branch (not `master`).

|  Status  | Feature  |  Assignee | Additional Details  |
|:--------:|:--------:|:---------:|:-------------------:|
| In-Progress  | Layer  | @maxwellbrown1990  |  |
|   | Tracing  |   | Most of the tracing implementation has been stubbed out. It should be re-implemented and tests added.   |
|   | Schedule  |   |  The signature of the `Schedule` datatype has been reworked in 2.x  |
|   | Scope  |   | The basic `Scope` has been implemented but it is feature incomplete  |
|   | Metrics  |   | The metric framework needs a second set of eyes - there's a lot of duplication. And it needs tests.   |
|   | Metrics  |   | The metric framework needs a second set of eyes - there's a lot of duplication. And it needs tests.   |
|   | XPure  |   | The `XPure` datatype needs to be integrated into the `FiberContext` run loop.   |
|   | Async  |   | The `Async` datatype should be reviewed to see if it is needed.  |
|   | STM Datatypes |   | There are many STM datatypes that are missing currently. |
|   | STM API |   | The API of STM should be audited against the ZIO 2.x branch. |
|   | Collections API |   | The naming of the functions in the collections directory should be unified with ZIO 2.x (i.e. `...M` functions should be `...Effect`, `...Unit` should be `...Discard`) |
|   | Stack Traces |   | In ZIO 2.x, error stacktraces closely mirror Java stacktraces - we should see if we can acoomplish the same or similar.  |

I will try to update the list above as more issues that need to be addressed are identified. If it is determined that the scope of this issue is more suited to a GitHub project, it can be enabled at a later date.

**Comment by patroza** [comment]
> `...M` functions should be `...Effect`

Would that be `...Effect`, `...Sync` etc?

**Comment by mikearnaldi** [comment]
@patroza yes that would be the change, I agree the `M` could stay in the compiler extension (maybe to some degree) even though I would try to find a better name for `M` given that `M` makes no sense to most

**Comment by patroza** [comment]
@mikearnaldi 
`M` - `Monad` - `Monadic`
`PM` - `Poly` - `Polymorphic`
;-)

**Comment by mikearnaldi** [comment]
something like `X` anyway I'll probably wait until I have some more concrete ideas about how to manage multiple overloads in extensions

**Comment by ax-hkightlinger** [comment]
Regarding `Async`, I would argue in favor of keeping it and am willing to spend more time developing / maintaining it. The 80 kb runtime for `Effect` is pretty tough to swallow compared to 3.5 kb for `Async` on the frontend.

**Comment by mikearnaldi** [comment]
@ax-hkightlinger I think we should keep it but extend both sync (that will be branched off from xpure) and async to support a failure Cause not a simple error like it is now so it becomes better interoperable

**Comment by ax-hkightlinger** [comment]
I agree. I'm happy to make an attempt at that work if you like to offload it. I can also try and add tracing.

**Comment by mikearnaldi** [comment]
> I agree. I'm happy to make an attempt at that work if you would like to offload it. I can also try and add tracing.

You're welcome to let me know if you need any help

**Comment by wesselvdv** [comment]
#1232 This PR should add the remaining STM data types.
